File: dot_to_xyflow.py
import re
import json
import sys

def parse_dot(dot_content):
    # Remove comments and global settings
    dot_content = re.sub(r'//.*', '', dot_content)
    dot_content = re.sub(r'rankdir=.*;', '', dot_content)
    
    nodes = {}
    edges = []
    
    # 1. Extract Explicit Nodes
    # Matches: ID [attr1="...", attr2=...]
    # We use a non-greedy match for attributes to avoid over-matching
    node_pattern = re.compile(r'^\s*(\w+)\s*\[([^\]]+)\]', re.MULTILINE)
    label_pattern = re.compile(r'label="([^"]+)"')
    shape_pattern = re.compile(r'shape=(\w+)')

    for match in node_pattern.finditer(dot_content):
        node_id, attr_string = match.groups()
        if node_id in ['node', 'graph', 'edge']: continue
        
        label_match = label_pattern.search(attr_string)
        label = label_match.group(1).replace('\\n', '\n') if label_match else node_id
        
        shape_match = shape_pattern.search(attr_string)
        shape = shape_match.group(1) if shape_match else "box"

        nodes[node_id] = {
            "id": node_id,
            "data": {"label": label},
            "position": {"x": 0, "y": 0},
            "type": "input" if "input" in node_id.lower() else ("output" if shape == "component" else "default"),
        }

    # 2. Extract Edges
    # Matches: Source -> Target [label="..."]
    edge_pattern = re.compile(r'(\w+)\s*->\s*(\w+)(?:\s*\[([^\]]+)\])?')
    for i, match in enumerate(edge_pattern.finditer(dot_content)):
        src, tgt, attr_string = match.groups()
        if src in ['node', 'graph', 'edge']: continue
        
        # Add nodes if they were used in edges but not explicitly defined
        if src not in nodes: nodes[src] = {"id": src, "data": {"label": src}, "position": {"x": 0, "y": 0}, "type": "default"}
        if tgt not in nodes: nodes[tgt] = {"id": tgt, "data": {"label": tgt}, "position": {"x": 0, "y": 0}, "type": "default"}

        edge = {
            "id": f"e{i}",
            "source": src,
            "target": tgt,
            "markerEnd": {"type": "arrowclosed"}
        }
        
        if attr_string:
            label_match = label_pattern.search(attr_string)
            if label_match:
                edge["label"] = label_match.group(1)
        
        edges.append(edge)

    return {"nodes": list(nodes.values()), "edges": edges}

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python dot_to_xyflow.py <filename.dot>")
        sys.exit(1)
        
    with open(sys.argv[1], 'r') as f:
        content = f.read()
        
    result = parse_dot(content)
    
    # Vertical layout (simpler for React Flow preview)
    for i, node in enumerate(result['nodes']):
        node['position'] = {"x": 250, "y": i * 120}

    print(json.dumps(result, indent=2))

File: dot_to_xyflow.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"strings"

	"github.com/goccy/go-graphviz"
)

type NodeData struct {
	Label string `json:"label"`
}

type Position struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type Node struct {
	ID       string   `json:"id"`
	Data     NodeData `json:"data"`
	Position Position `json:"position"`
	Type     string   `json:"type"`
}

type MarkerEnd struct {
	Type string `json:"type"`
}

type Edge struct {
	ID        string    `json:"id"`
	Source    string    `json:"source"`
	Target    string    `json:"target"`
	MarkerEnd MarkerEnd `json:"markerEnd"`
	Label     string    `json:"label,omitempty"`
}

type XYFlowData struct {
	Nodes []Node `json:"nodes"`
	Edges []Edge `json:"edges"`
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run dot_to_xyflow.go <filename.dot>")
		return
	}

	content, err := os.ReadFile(os.Args[1])
	if err != nil {
		log.Fatalf("Error reading file: %v", err)
	}

	ctx := context.Background()
	gv, err := graphviz.New(ctx)
	if err != nil {
		log.Fatalf("Error creating graphviz context: %v", err)
	}
	defer gv.Close()

	graph, err := graphviz.ParseBytes(content)
	if err != nil {
		log.Fatalf("Error parsing DOT: %v", err)
	}
	defer graph.Close()

	// Set layout engine (default to dot)
	gv.SetLayout(graphviz.DOT)

	// Render to XDOT to trigger layout and populate attributes
	if err := gv.Render(ctx, graph, graphviz.XDOT, io.Discard); err != nil {
		log.Fatalf("Error performing layout: %v", err)
	}

	// Get graph bounding box to flip Y axis
	bb := graph.GetStr("bb")
	var llx, lly, urx, ury float64
	fmt.Sscanf(bb, "%f,%f,%f,%f", &llx, &lly, &urx, &ury)
	graphHeight := ury

	nodes := []Node{}
	edges := []Edge{}

	// Iterate nodes
	n, err := graph.FirstNode()
	for n != nil && err == nil {
		id, _ := n.Name()
		label := n.GetStr("label")
		if label == "" {
			label = id
		}
		label = strings.ReplaceAll(label, "\n", "\n")

		posStr := n.GetStr("pos")
		var x, y float64
		if posStr != "" {
			fmt.Sscanf(posStr, "%f,%f", &x, &y)
		}

		// Graphviz pos is center. XYFlow is top-left.
		// width/height are in inches, convert to points (72 dpi)
		width, _ := strconv.ParseFloat(n.GetStr("width"), 64)
		height, _ := strconv.ParseFloat(n.GetStr("height"), 64)
		wPts := width * 72
		hPts := height * 72
		
		// Adjust x from center to left
		x = x - (wPts / 2)
		// Flip Y: Graphviz is bottom-up, XYFlow is top-down
		// y is center, so top-left Y in XYFlow is graphHeight - (y + hPts/2)
		y = graphHeight - (y + hPts/2)

		nodeType := "default"
		shape := n.GetStr("shape")
		if shape == "component" || shape == "doublecircle" {
			nodeType = "output"
		} else if shape == "note" || shape == "invhouse" || strings.Contains(strings.ToLower(id), "input") {
			nodeType = "input"
		}

		nodes = append(nodes, Node{
			ID: id,
			Data: NodeData{Label: label},
			Position: Position{X: x, Y: y},
			Type: nodeType,
		})
		n, err = graph.NextNode(n)
	}

	// Iterate edges
	edgeCount := 0
	n, err = graph.FirstNode()
	for n != nil && err == nil {
		e, err := graph.FirstOut(n)
		for e != nil && err == nil {
			tail, _ := e.Tail()
			head, _ := e.Head()
			src, _ := tail.Name()
			tgt, _ := head.Name()
			
			label := e.GetStr("label")

			edges = append(edges, Edge{
				ID: fmt.Sprintf("e%d", edgeCount),
				Source: src,
				Target: tgt,
				MarkerEnd: MarkerEnd{Type: "arrowclosed"},
				Label: label,
			})
			edgeCount++
			e, err = graph.NextOut(e)
		}
		n, err = graph.NextNode(n)
	}

	out, _ := json.MarshalIndent(XYFlowData{Nodes: nodes, Edges: edges}, "", "  ")
	fmt.Println(string(out))
}
